Fix importer and morph with HWIntrinsic SIMD types

fiigii opened the pull request:
Fix https://github.com/dotnet/coreclr/issues/16236

Ideally, this bug should be fixed by unifying `isSIMDClass` for `S.N.Vector` and hardware-intrinsic vectors. However, we just moved `Vector<T>` into mscrolib, and `Vector2/3/4` are still in the individual assembly. So, in the future, if we decide to move `Vector2/3/4` into mscrolib as well, all the `S.N.Vector` can be recognized by `[Intrinsic]` mechanism and `isSIMDClass`/`isHWSIMDClass` can be combined.

fiigii commented:
After apply this change, another failure appears under jitstress tests.
```

Assert failure(PID 19736 [0x00004d18], Thread: 18680 [0x48f8]): Assertion failed 'child->isContained()' in 'JIT.HardwareIntrinsics.X86.SimpleBinaryOpTest__AddScalarSingle:RunReflectionScenario_LoadAligned():this' (IL size 180)

    File: d:\workspace\coreclr-vector\src\jit\lsraxarch.cpp Line: 411
    Image: D:\workspace\coreclr-vector\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe
```

This is cased by `GT_ADDR`'s child (`LoadAlignedVector128`) not marked as "contained" in the test of  `Sse.AddScalar`.
```
    [ 6] 101 (0x065) box 1B000007
Compiler::impImportAndPushBox -- handling BOX(value class) via helper call because: optimizing for size
    [ 6]  82 (0x052) stelem.ref

               [000098] ------------              *  STMT      void  (IL   ???...  ???)
               [000097] --CXG-------              \--*  CALL help void   HELPER.CORINFO_HELP_ARRADDR_ST
               [000082] ------------ arg0            +--*  LCL_VAR   ref    V08 tmp6         
               [000083] ------------ arg1            +--*  CNS_INT   int    0
               [000093] --CXG------- arg2            \--*  CALL help ref    HELPER.CORINFO_HELP_BOX
               [000089] ------------ arg0               +--*  CNS_INT(h) long   0x7ffd2246bfd0 class
               [000090] --CXG------- arg1               \--*  ADDR      byref // This node's child should be contained
               [000088] --CXG-------                       \--*  HWIntrinsic simd16 float LoadAlignedVector128
               [000087] --CXG-------                          \--*  CALL      long   JIT.HardwareIntrinsics.X86.SimpleBinaryOpTest__DataTable`1[Single][System.Single].get_inArray1Ptr
               [000086] ---XG------- this in rcx                 \--*  ADDR      byref 
               [000085] ---XG-------                                \--*  FIELD     struct _dataTable
               [000084] ------------                                   \--*  LCL_VAR   ref    V02 tmp0  
```

```csharp
        public void RunReflectionScenario_LoadAligned()
        {
            var result = typeof(Sse).GetMethod(nameof(Sse.AddScalar), new Type[] { typeof(Vector128<Single>), typeof(Vector128<Single>) })
                                     .Invoke(null, new object[] {
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray1Ptr)),
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray2Ptr))
                                     });

            Unsafe.Write(_dataTable.outArrayPtr, (Vector128<Single>)(result));
            ValidateResult(_dataTable.inArray1Ptr, _dataTable.inArray2Ptr, _dataTable.outArrayPtr);
        }
```
  
  
  
However, the test of `Sse.Add` is compiled to a different IR shape, and it works fine.
```
    [ 6] 101 (0x065) box 1B000007
Compiler::impImportAndPushBox -- handling BOX(value class) via inline allocate/copy sequence

lvaGrabTemp returning 7 (V07 tmp5) called for Single-def Box Helper.

lvaSetClass: setting class for V07 to (00007FFD2246BFD0) System.Runtime.Intrinsics.Vector128`1[Single]  [exact]


               [000095] ------------              *  STMT      void  (IL   ???...  ???)
               [000092] ------------              |  /--*  ALLOCOBJ  ref   
               [000091] ------------              |  |  \--*  CNS_INT(h) long   0x7ffd2246bfd0 class
               [000094] -A----------              \--*  ASG       ref   
               [000093] D------N----                 \--*  LCL_VAR   ref    V07 tmp5         


               [000101] ------------              *  STMT      void  (IL   ???...  ???)
               [000090] --C---------              |  /--*  HWIntrinsic simd16 float LoadAlignedVector128
               [000089] --C---------              |  |  \--*  RET_EXPR  long  (inl return from call [000087])
               [000100] -AC-----R---              \--*  ASG       simd16 (copy)
               [000099] ------------                 \--*  BLK(16)   simd16
               [000097] ------------                    |  /--*  CNS_INT   long   8
               [000098] ------------                    \--*  ADD       byref 
               [000096] ------------                       \--*  LCL_VAR   ref    V07 tmp5         

    [ 6] 106 (0x06a) stelem.ref

               [000108] ------------              *  STMT      void  (IL   ???...  ???)
               [000107] --CXG-------              \--*  CALL help void   HELPER.CORINFO_HELP_ARRADDR_ST
               [000082] ------------ arg0            +--*  LCL_VAR   ref    V04 tmp2         
               [000083] ------------ arg1            +--*  CNS_INT   int    1
               [000103] ------------ arg2            \--*  BOX       ref   
               [000102] ------------                    \--*  LCL_VAR   ref    V07 tmp5 
```

```csharp
        public void RunReflectionScenario_LoadAligned()
        {
            var result = typeof(Sse).GetMethod(nameof(Sse.Add), new Type[] { typeof(Vector128<Single>), typeof(Vector128<Single>) })
                                     .Invoke(null, new object[] {
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray1Ptr)),
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray2Ptr))
                                     });

            Unsafe.Write(_dataTable.outArrayPtr, (Vector128<Single>)(result));
            ValidateResult(_dataTable.inArray1Ptr, _dataTable.inArray2Ptr, _dataTable.outArrayPtr);
        }
```

Investigaing in progress. @CarolEidt @tannergooding @AndyAyersMS @BruceForstall PTAL

fiigii commented:
However, the test of `Sse.Add` is compiled to a different IR shape, and it works fine.
```
    [ 6] 101 (0x065) box 1B000007
Compiler::impImportAndPushBox -- handling BOX(value class) via inline allocate/copy sequence

lvaGrabTemp returning 7 (V07 tmp5) called for Single-def Box Helper.

lvaSetClass: setting class for V07 to (00007FFD2246BFD0) System.Runtime.Intrinsics.Vector128`1[Single]  [exact]


               [000095] ------------              *  STMT      void  (IL   ???...  ???)
               [000092] ------------              |  /--*  ALLOCOBJ  ref   
               [000091] ------------              |  |  \--*  CNS_INT(h) long   0x7ffd2246bfd0 class
               [000094] -A----------              \--*  ASG       ref   
               [000093] D------N----                 \--*  LCL_VAR   ref    V07 tmp5         


               [000101] ------------              *  STMT      void  (IL   ???...  ???)
               [000090] --C---------              |  /--*  HWIntrinsic simd16 float LoadAlignedVector128
               [000089] --C---------              |  |  \--*  RET_EXPR  long  (inl return from call [000087])
               [000100] -AC-----R---              \--*  ASG       simd16 (copy)
               [000099] ------------                 \--*  BLK(16)   simd16
               [000097] ------------                    |  /--*  CNS_INT   long   8
               [000098] ------------                    \--*  ADD       byref 
               [000096] ------------                       \--*  LCL_VAR   ref    V07 tmp5         

    [ 6] 106 (0x06a) stelem.ref

               [000108] ------------              *  STMT      void  (IL   ???...  ???)
               [000107] --CXG-------              \--*  CALL help void   HELPER.CORINFO_HELP_ARRADDR_ST
               [000082] ------------ arg0            +--*  LCL_VAR   ref    V04 tmp2         
               [000083] ------------ arg1            +--*  CNS_INT   int    1
               [000103] ------------ arg2            \--*  BOX       ref   
               [000102] ------------                    \--*  LCL_VAR   ref    V07 tmp5 
```

```csharp
        public void RunReflectionScenario_LoadAligned()
        {
            var result = typeof(Sse).GetMethod(nameof(Sse.Add), new Type[] { typeof(Vector128<Single>), typeof(Vector128<Single>) })
                                     .Invoke(null, new object[] {
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray1Ptr)),
                                        Sse.LoadAlignedVector128((Single*)(_dataTable.inArray2Ptr))
                                     });

            Unsafe.Write(_dataTable.outArrayPtr, (Vector128<Single>)(result));
            ValidateResult(_dataTable.inArray1Ptr, _dataTable.inArray2Ptr, _dataTable.outArrayPtr);
        }
```

Investigaing in progress. @CarolEidt @tannergooding @AndyAyersMS @BruceForstall PTAL

fiigii commented:
test Windows_NT jitstress2

fiigii commented:
test Windows_NT Checked jitstress2 Build and Test

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test

tannergooding commented:
@fiigii, for the `AddScalar` test, it should not be marked as contained (the operation is a m32, while the `LoadAligned` is a m128).

I wonder if the failure is also due to a disconnect in expected vs actual memory size.

tannergooding commented on src/jit/importer.cpp:
```diff
@@ -18277,7 +18277,11 @@ void Compiler::impInlineInitVars(InlineInfo* pInlineInfo)
         // But we only need to normalize it if it is a TYP_STRUCT
         // (which we need to do even if we have already set foundSIMDType).
-        if ((!foundSIMDType || (sigType == TYP_STRUCT)) && isSIMDClass(&(lclVarInfo[0].lclVerTypeInfo)))
+        if ((!foundSIMDType || (sigType == TYP_STRUCT)) && (isSIMDClass(&(lclVarInfo[0].lclVerTypeInfo))
```
Why can't the new logic in `isHWSIMDClass` just be added to `isSIMDClass`?

fiigii replied:
Because `isSIMDClass` is used for `S.N.Vector` recognition. https://github.com/dotnet/coreclr/blob/master/src/jit/simd.cpp#L137

fiigii commented:
> I wonder if the failure is also due to a disconnect in expected vs actual memory size.

Probably not, here Load's parent node is `GT_ADDR`. And the child of `GT_ADDR` must be contained.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked GCStress=4 Build and Test

fiigii commented:
@BruceForstall could you help me to trigger the GCStress tests that you detected https://github.com/dotnet/coreclr/issues/16370?

BruceForstall commented:
The CI trigger phrase is "... Windows_NT x86 Checked gcstress0xc". You can run the tests locally by setting `COMPlus_GCStress=C` and running the one test, or using
```
 test\runtest x86 checked gcstresslevel C
```

You can replace "x86" with "x64".

fiigii commented:
test Windows_NT x86 Checked gcstress0xc
test Windows_NT x64 Checked gcstress0xc

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT x86 Checked jitstress2 Build and Test

fiigii commented:
Sometimes, the compiler may generate `GT_ADDR(GT_HWIntrinsic)` as arguments for call-by-ref to functions/helper. The latest commit breaks this kind of nodes and inserts temp-variable/block-copy.
With this change, all jitstress tests passed on my local machine.

@CarolEidt @tannergooding PTAL

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

fiigii commented:
The  jitstress failures are unrelated.

fiigii commented:
@AndyAyersMS Could you take a look at this change as well? One of the issues that the PR fixed is from `impImportAndPushBox` with MinOpt https://github.com/dotnet/coreclr/pull/16365#issuecomment-365442725. As I know, you are familar with this part.

AndyAyersMS commented:
Given that you're introducing the same 3 or so patterns at various points in the code it would be nice to encapsulate these as utilities to make the code more readable -- especially with the need for ifdefs -- and perhaps save a bit on subexpression eval costs. For instance:
```C++
(isSIMDClass(&(lclVarInfo[i].lclVerTypeInfo))
+#ifdef FEATURE_HW_INTRINSICS
+             || isHWSIMDClass(&(lclVarInfo[i].lclVerTypeInfo))
+#endif
```
could be repackaged as (with you choosing a better name)
```C++
isSIMDClassOrHWSIMDClass(...)
```
likewise with
```C++
    if (src->OperGet() == GT_SIMD || src->OperIsSimdHWIntrinsic())
```
By the way the subtle difference between
```C++
src->OperIsSimdHWIntrinsic()
```
and
```C++
src->OperGet() == GT_HWIntrinsic
```
is annoying and seems error prone, and I think you may have gotten it wrong here in at least one spot.

fiigii commented:
@tannergooding thank you so much for the suggestions. Will fix.

fiigii commented:
Addressed feedback.

@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT x86 Checked jitstress2 Build and Test

tannergooding commented on src/jit/morph.cpp:
```diff
@@ -10223,7 +10227,11 @@ GenTree* Compiler::fgMorphBlockOperand(GenTree* tree, var_types asgType, unsigne
             if ((indirTree != nullptr) && (lclNode == nullptr) && (indirTree->Addr()->OperGet() == GT_ADDR) &&
-                (indirTree->Addr()->gtGetOp1()->gtOper == GT_SIMD))
+                (indirTree->Addr()->gtGetOp1()->gtOper == GT_SIMD
+#ifdef FEATURE_HW_INTRINSICS
```
Why is this one ifdef'd, but not the others?

Also, should we have a helper method for `OperIsSimd || OperIsSimdHWIntrinsic`?

fiigii replied:
> Why is this one ifdef'd, but not the others?

Ah, yes, we can remove ifdef.

> should we have a helper method for OperIsSimd || OperIsSimdHWIntrinsic?

Good point. Will do.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT x86 Checked jitstress2 Build and Test

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT x86 Checked jitstress2 Build and Test

tannergooding commented on src/jit/morph.cpp:
```diff
@@ -11378,6 +11378,15 @@ GenTree* Compiler::getSIMDStructFromField(GenTree*   tree,
                 *pBaseTypeOut         = simdNode->gtSIMDBaseType;
             }
+#ifdef FEATURE_HW_INTRINSICS
+            else if (obj->OperIsSimdHWIntrinsic())
```
Should we merge this with the `GT_SIMD` check above (given that `GenTreeHWIntrinsic` and `GenTreeSIMD` both inherit from `GenTreeJitIntrinsic`, which is the node that actually carries this data)?

fiigii replied:
> inherit from GenTreeJitIntrinsic, which is the node that actually carries this data)?

Yes, but we do not have `AsJitIntrinsic`.

tannergooding replied:
Might be worth investigating if its worthwhile to add later (maybe as part of https://github.com/dotnet/coreclr/issues/15641)

fiigii replied:
> if its worthwhile to add later (maybe as part of #15641)

Agree.

briansull commented on src/jit/morph.cpp:
```diff
@@ -9306,17 +9306,17 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
     CORINFO_CLASS_HANDLE clsHnd = NO_CLASS_HANDLE;
 #ifdef FEATURE_SIMD
-    // importer introduces cpblk nodes with src = GT_ADDR(GT_SIMD)
+    // importer introduces cpblk nodes with src = GT_ADDR(GT_SIMD/GW_HWIntrinsic)
```
GW_HWIntrinsic => GT_HWIntrinsic

briansull commented on src/jit/morph.cpp:
```diff
@@ -9306,17 +9306,17 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
     // copyblk appropriately. Otherwise, the transformation made in this
     // routine will prevent rationalizer logic and we might end up with
-    // GT_ADDR(GT_SIMD) node post rationalization, leading to a noway assert
+    // GT_ADDR(GT_SIMD/GW_HWIntrinsic) node post rationalization, leading to a noway assert
```
GW_HWIntrinsic => GT_HWIntrinsic

fiigii commented on src/jit/compiler.h:
```diff
@@ -7634,6 +7634,38 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+        if (isIntrinsicType(clsHnd))
+        {
+            const char* namespaceName = nullptr;
+            (void)getClassNameFromMetadata(clsHnd, &namespaceName);
```
> getClassNameFromMetadata is going to be extremely slow

Really? There is another similar method `getMethodNameFromMetadata` is used in "commonly executed code paths" as well. And I remember @AndyAyersMS said the JIT-EE interfaces have been optimized (retrieval strings from metadata).

fiigii replied:
Is this method much slower than `isInSIMDModule`?

AndyAyersMS replied:
@briansull `getClassNameFromMetadata` should be fairly fast -- it does not do any formatting or allocation. We are relying on it for the new style intrinsic recognition.

AndyAyersMS replied:
Also we're only calling it selectively here, on types that are potential intrinsic types. So it is pay for play.

briansull replied:
Getting the name from metadata is somewhat slow, but on top of that there is a the string compare is is also slow.   
`strcmp(namespaceName, "System.Runtime.Intrinsics") == 0;`
I think that a cache of the last clsHnd value that passed the strcmp test would be best.

fiigii replied:
> I think that a cache of the last clsHnd value that passed the strcmp test would be best.

Good idea. But here is not the only one "checking class/namespace name" in the common code path, and I think we need to optimize `S.N.Vector` recognition at first https://github.com/dotnet/coreclr/pull/16365#issue-168983511. 
Can we consider this optimization later?

briansull replied:
Sure, We can consider this optimization for the future.
I was never blocking this checkin on this issue.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT x86 Checked jitstress2 Build and Test

fiigii commented:
All tests passed (one timeout), shall we merge this PR to unblock other test work?
